components/Reader/Reader.js

```diff
@@ -8,71 +8,234 @@
+            onRequestPrev={this.props.onRequestPrevious}
+            disableNext={disableNext}
+            disablePrev={disablePrev}
+          />
+          <Card>
+            <Toolbar
+              showOnMobile={this.state.showOverlay}
+              onRequestClose={this.props.onRequestClose}
+            />
+            <Box px={[40, 120]} pb={20} flex="1 0 auto">
+              <BookMeta
+                currentChapter={chapterNumber}
+                totalChapters={numOfChapters}
+                hideOnTablet
+              />
+              {chapter && (
+                <Page dangerouslySetInnerHTML={createMarkup(chapter)} />
+              )}
+              <BookMeta
+                currentChapter={chapterNumber}
+                totalChapters={numOfChapters}
+                title={book.title}
+              />
+            </Box>
+          </Card>
+        </Swipeable>
+        <KeyDown
+          when="ArrowRight"
+          then={this.props.onRequestNext}
+          disabled={disableNext}
+        />
+        <KeyDown
+          when="ArrowLeft"
+          then={this.props.onRequestPrevious}
+          disabled={disablePrev}
+        />
+      </Container>
+    );
   }
+}
+
+type ReaderContainerState = {
+  chapters: { [number]: Chapter },
+  chapter: number,
+};
+
+type ReaderContainerProps = {
+  book: Book,
+  initialChapter: ?string,
+};
+
+export default class ReaderContainer extends React.Component<
+  ReaderContainerProps,
+  ReaderContainerState,
+> {
+  constructor(props: ReaderContainerProps) {
+    super(props);
 
-  componentWillReceiveProps(nextProps: Props) {
-    if (nextProps.chapter !== this.props.chapter) {
-      this.loadChapter(nextProps.chapter);
+    // Convert the chapter in the url to an int and make sure it is in a valid chapter range (or be set to 1)
+    let initialChapter = parseInt(props.initialChapter, 10) || 1;
+    if (initialChapter < 1 || initialChapter > props.book.chapters.length) {
+      initialChapter = 1;
     }
+
+    this.state = {
+      chapters: {},
+      chapter: initialChapter,
+    };
+  }
+
+  componentDidMount() {
+    this.loadChapter(this.state.chapter);
+    // Load the next chapter
+    this.loadChapter(this.state.chapter + 1);
   }
 
+  // Go back to the book details when closing the reader
+  onRequestClose = () => {
+    Router.replaceRoute('book', {
+      id: this.props.book.id,
+      lang: this.props.book.language.code,
+    });
+  };
+
+  onRequestNext = () => {
+    if (this.state.chapter < this.props.book.chapters.length) {
+      this.loadChapter(this.state.chapter + 1);
+      this.loadChapter(this.state.chapter + 2);
+      this.setState(
+        state => ({ chapter: state.chapter + 1 }),
+        this.changeChapterInUrl,
+      );
+    }
+  };
+
+  onRequestPrevious = () => {
+    if (this.state.chapter > 1) {
+      this.loadChapter(this.state.chapter - 1);
+      this.loadChapter(this.state.chapter - 2);
```

Comment:
Første kapittel har id 0?

Reply:
Yes, but "Chapter 0 / 12" doesn't look very pretty unless you're an engineer, so we subtract by one in the `loadChapter` method so we lookup correctly in the array there. In other words, we use one indexed chapters.